docs/RFCS/20190318_error_handling.md

```diff
@@ -0,0 +1,2090 @@
+  - [Suspicious comparisons of the error object](#Suspicious-comparisons-of-the-error-object)
+  - [Suspicious assertions on the error type](#Suspicious-assertions-on-the-error-type)
+  - [Suspicious error predicates](#Suspicious-error-predicates)
+  - [Suspicious comparisons of the error message](#Suspicious-comparisons-of-the-error-message)
+  - [Suspicious reliance on the pg error code](#Suspicious-reliance-on-the-pg-error-code)
+  - [Suspicious flattening of errors](#Suspicious-flattening-of-errors)
+- [Error handling outside of CockroachDB](#Error-handling-outside-of-CockroachDB)
+  - [Go error handling pre-v2](#Go-error-handling-pre-v2)
+  - [`github.com/pkg/errors`](#githubcompkgerrors)
+  - [`github.com/hashicorp/errwrap`](#githubcomhashicorperrwrap)
+  - [`upspin.io/errors`](#upspinioerrors)
+  - [Go error handling v2+](#Go-error-handling-v2-)
+
+
+## Current error handling in CockroachDB
+
+### Error types in use
+
+CockroachDB currently uses:
+
+- the fundamental `errorString` instantiated via `errors.New` (from `golang.org/pkg/errors`) and `fmt.Errorf`
+- the various error types from `github.com/pkg/errors`: `fundamental`, `withMessage` and `withStack`
+- `roachpb.Error`, which may include any of the other `roachpb` errors via its `ErrorDetail` field
+- `pgerror.Error` which can encode pg codes, detail, hints, safe details, telemetry key, etc
+- `distsqlpb.Error` which can wire-encode either some of the types in `roachpb` or a `pgerror.Error`.
+
+### Error protocols
+
+CockroachDB currently uses all of the [unreliable 4
+"standard"/idiomatic
+methods](#Unreliable-standard-ways-to-inspect-errors) to inspect
+errors:
+
+- comparison with reference errors, used both for standard library
+  errors (`io.EOF`, `context.Canceled` etc) and also
+  CockroachDB-specific reference errors (`sql.errNoZoneConfigApplies`,
+  `grpcutil.ErrCannotReuseClientConn`, etc).
+
+  See [Suspicious comparisons of the error
+  object](#Suspicious-comparisons-of-the-error-object) below.
+
+- type assertions to known error types, mostly to `roachpb` errors and
+  `*pgerror.Error`.
+
+  See [Suspicious assertions on the error
+  type](#Suspicious-assertions-on-the-error-type) below.
+
+- error predicates, for example `scrub.IsScrubError()`,
+  `sqlbase.IsQueryCanceledError()`.
+
+  See [Suspicious error predicates](#Suspicious-error-predicates)
+  below.
+
+- comparisons on the message string, for example `strings.Contains(err.Error(), "must be after replica GC threshold")`.
+
+  See [Suspicious comparisons of the error
+  message](#Suspicious-comparisons-of-the-error-message) below.
+
+It also sometimes (more rarely) depends on the pg error code to
+determine further logic, for example `if pgErr.Code ==
+pgerror.CodeUndefinedColumnError`.
+
+See [Suspicious reliance on the pg error
+code](#Suspicious-reliance-on-the-pg-error-code) below.
+
+
+## Problematic error use cases
+
+### Suspicious comparisons of the error object
```

Comment:
Comparing to `io.EOF` isn't suspicious most of the time - that's how Go recommends that you check for error conditions from its library calls, which I'm sure is done in these code samples, right?